exec: add multi column equality to vectorized mergejoiner

georgeutsin opened the pull request:
Added the ability to join on multiple columns in the vectorized
merge joiner. This was done by inverting the previous method of
generating groups; instead of assuming no matches and discovering
matches group by group, if we assume that everything is a match
and filter out the rows that don't match in the join, we can use
multiple columns in this approach.

This PR also includes some work on output shaping, to ensure that
we still have decently shaped outputs, even when a batch ends on
a small run. Also fixed the benchmarks in this PR.

Updated benchmarks for the mergejoiner:

```
BenchmarkMergeJoiner/rows=1024-8         	   20000	     67231 ns/op	 974.77 MB/s
BenchmarkMergeJoiner/rows=4096-8         	    5000	    257487 ns/op	1018.08 MB/s
BenchmarkMergeJoiner/rows=16384-8        	    2000	    995437 ns/op	1053.38 MB/s
BenchmarkMergeJoiner/rows=1048576-8      	      20	  62822638 ns/op	1068.23 MB/s
BenchmarkMergeJoiner/oneSideRepeat-rows=1024-8         	   20000	     67343 ns/op	 973.16 MB/s
BenchmarkMergeJoiner/oneSideRepeat-rows=4096-8         	    5000	    256988 ns/op	1020.06 MB/s
BenchmarkMergeJoiner/oneSideRepeat-rows=16384-8        	    2000	    997690 ns/op	1051.00 MB/s
BenchmarkMergeJoiner/oneSideRepeat-rows=1048576-8      	      20	  62184741 ns/op	1079.19 MB/s
BenchmarkMergeJoiner/bothSidesRepeat-rows=1024-8       	   20000	     68672 ns/op	 954.32 MB/s
BenchmarkMergeJoiner/bothSidesRepeat-rows=4096-8       	    1000	   1336630 ns/op	 196.12 MB/s
BenchmarkMergeJoiner/bothSidesRepeat-rows=16384-8      	     100	  18478767 ns/op	  56.74 MB/s
BenchmarkMergeJoiner/bothSidesRepeat-rows=32768-8      	      20	  80351272 ns/op	  26.10 MB/s
```

Note that the performance degradation in the 'bothSidesRepeat'
is actually due to incorrect benchmarking in the first place.
Regardless, this is still an area of investigation.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35655)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Very nice work! It seems simpler than I thought it would be.

I think reusing `ring.Buffer` for the circular buffer you're using can make it even easier to follow.

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-:-L_nRx1T7Q0B0hk6huWw:b-m7s0tm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @georgeutsin, and @jordanlewis)

___
*[pkg/sql/exec/mergejoiner.go, line 145 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nIof-1HQXvSo3WUsX:-L_nIof-1HQXvSo3WUsY:b-8ymxdh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L145)):*
> ```Go
> }
> 
> // getBatch takes a side as input and returns either the next batch (from source),
> ```

[nit]: it is no longer "a side".
___
*[pkg/sql/exec/mergejoiner.go, line 241 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nKgIP08Ad02nSEbjV:-L_nKgIP08Ad02nSEbjW:b3ze9gu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L241)):*
> ```Go
> // getExpectedOutCount is a helper function to generate the right length of output,
> // if there are no output columns. ex: in the case of a COUNT().
> func (c *mergeJoinOp) getExpectedOutCount(groups []group, groupsLen int, outCount uint16) uint16 {
> ```

[nit]: I don't immediately see what `outCount` stands for and where it is coming from. Please extend the comment to explain it.
___
*[pkg/sql/exec/mergejoiner.go, line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nKDhGB2sthDwLp-pj:-L_nKDhGB2sthDwLp-pk:b-m62q7k) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L258)):*
> ```Go
> }
> 
> // completeRun takes the existing run in state and extends it given the source input.
> ```

[nit]: I think usually `takes` refers to the input argument, but the run comes from the state. I'd probably say something like `completeRun given the source input extends the existing run in state`.
___
*[pkg/sql/exec/mergejoiner.go, line 265 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nJN301znhcmm2KwFL:-L_nJN301znhcmm2KwFM:bxpsza2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L265)):*
> ```Go
> func (c *mergeJoinOp) completeRun(
> 	source *mergeJoinInput, bat ColBatch, rowIdx int,
> ) (int, ColBatch, int, []uint16, []int64) {
> ```

[nit]: with this many return values consider naming them.
___
*[pkg/sql/exec/mergejoiner.go, line 278 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nLf--C8LD6JX_ISdx:-L_nLf--C8LD6JX_ISdy:b3urdte) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L278)):*
> ```Go
> 	}
> 
> 	// Check all equality columns before the last minor equality column to make sure we're in the same run.
> ```

[nit]: what does `minor` mean in this context?
___
*[pkg/sql/exec/mergejoiner.go, line 283 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nOcFS-65El9OBdgHv:-L_nOcFS-65El9OBdgHw:b-39nzye) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L283)):*
> ```Go
> 		prevVal := getValForIdx(savedRun.ColVec(int(colIdx)).Int64(), *savedRunIdx-1, nil)
> 		curVal := getValForIdx(bat.ColVec(int(colIdx)).Int64(), rowIdx, sel)
> 		if prevVal != curVal {
> ```

I think this is why you were asking about `tuplesDiffer`? If you want to use it, you could write something like (if `sel` is nil):
```
differ := false
for i := 0; i < len(source.eqCols)-1; i++ {
  colIdx := source.eqCols[i]
  tuplesDiffer(types.Int64, savedRun.ColVec(int(colIdx)), *savedRunIdx-1, bat.ColVec(int(colIdx)), rowIdx, &differ)
}
if differ {
  return rowIdx, bat, length, sel, keys
```
(But I guess you still need to template the rest, so it might not make sense to use it at the moment.)
___
*[pkg/sql/exec/mergejoiner.go, line 404 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nQiSREhvNWJjHqU46:-L_nQiSREhvNWJjHqU47:bjapa5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L404)):*
> ```Go
> 		}
> 
> 		// Phase 1: probe.
> ```

[nit]: I think this phase deserves at least a high level comment explaining what we're doing here.
___
*[pkg/sql/exec/mergejoiner.go, line 405 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nRGROCH6SDQV1VyGG:-L_nRGROCH6SDQV1VyGH:b8i6twk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L405)):*
> ```Go
> 
> 		// Phase 1: probe.
> 		bufferCap := len(c.leftGroups)
> ```

I think you're using a circular buffer here, and we have an implementation of such already in `ring.Buffer`. Consider using that if possible.
___
*[pkg/sql/exec/mergejoiner.go, line 416 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nQzF1BPQTu-Cu2kZg:-L_nQzF1BPQTu-Cu2kZh:b-istnte) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L416)):*
> ```Go
> 			rKeys = rBat.ColVec(int(c.right.eqCols[eqColIdx])).Int64()
> 			oldBufferEndIdx := bufferEndIdx
> 			// Iterate over all the groups in a column.
> ```

[nit]: `in the column`.
___
*[pkg/sql/exec/mergejoiner_test.go, line 490 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nQC3i9vJhmioqc6J8:-L_nQC3i9vJhmioqc6J9:b35xxjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner_test.go#L490)):*
> ```Go
> 	}
> }
> ```

Consider adding a randomized test as well with variable number of equality columns and small number of tuples. I think a randomized test I wrote for sort chunks might be a good starting point.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 148 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nN_es8fLipHZEP_e5:-L_nN_es8fLipHZEP_e6:b-qmj4j1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner_tmpl.go#L148)):*
> ```Go
> 	sel []uint16,
> 	destStartIdx uint16,
> ) (uint16, int) {
> ```

[nit]: it'd be useful to explain the return values in the comment.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nNvyb8Behwuwg2nsO:-L_nNvyb8Behwuwg2nsP:bqvu5gx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner_tmpl.go#L150)):*
> ```Go
> ) (uint16, int) {
> 	savedOutCount := 0
> 	outCount := uint16(0)
> ```

[nit]: this `outCount` seems to be not used anymore, maybe get rid off it? Or declare it within the for loop?


<!-- Sent from Reviewable.io -->

georgeutsin committed:
exec: add multi column equality to vectorized mergejoiner

Added the ability to join on multiple columns in the vectorized
merge joiner. This was done by inverting the previous method of
generating groups; instead of assuming no matches and discovering
matches group by group, if we assume that everything is a match
and filter out the rows that don't match in the join, we can use
multiple columns in this approach.

This PR also includes some work on output shaping, to ensure that
we still have decently shaped outputs, even when a batch ends on
a small run. Also fixed the benchmarks in this PR.

Updated benchmarks for the mergejoiner:

```
BenchmarkMergeJoiner/rows=1024-8                   20000             77494 ns/op         845.68 MB/s
BenchmarkMergeJoiner/rows=4096-8                    5000            296581 ns/op         883.88 MB/s
BenchmarkMergeJoiner/rows=16384-8                   2000           1161530 ns/op         902.75 MB/s
BenchmarkMergeJoiner/rows=1048576-8                   20          72865635 ns/op         920.99 MB/s
BenchmarkMergeJoiner/oneSideRepeat-rows=1024-8             20000             77621 ns/op         844.30 MB/s
BenchmarkMergeJoiner/oneSideRepeat-rows=4096-8              5000            294507 ns/op         890.11 MB/s
BenchmarkMergeJoiner/oneSideRepeat-rows=16384-8             2000           1153490 ns/op         909.05 MB/s
BenchmarkMergeJoiner/oneSideRepeat-rows=1048576-8             20          74413692 ns/op         901.83 MB/s
BenchmarkMergeJoiner/bothSidesRepeat-rows=1024-8           20000             79524 ns/op         824.10 MB/s
BenchmarkMergeJoiner/bothSidesRepeat-rows=4096-8            1000           1353760 ns/op         193.64 MB/s
BenchmarkMergeJoiner/bothSidesRepeat-rows=16384-8            100          18884081 ns/op          55.53 MB/s
BenchmarkMergeJoiner/bothSidesRepeat-rows=32768-8             20          82696516 ns/op          25.36 MB/s
```

Note that the performance degradation in the 'bothSidesRepeat'
is actually due to incorrect benchmarking in the first place.
Regardless, this is still an area of investigation.

Release note: None

Release note: None

asubiotto reviewed (changes requested):
Nice!

When sharing benchmark changes, it's usually easier to run `./scripts/bench` to get more reliable differences (the tool runs benchmarks multiple times to reduce noise and then uses `benchstat` to provide more reliable deltas). You can run it like `BENCHES=BenchmarkMergeJoiner PKG=./pkg/sql/exec BENCHTIMEOUT=5h ./scripts/bench master <your_branch>`. Make sure to commit your changes before running the command.

> Reviewed 3 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-:-L_wRd2K-xKqh880fZW0:buw0341)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/exec/mergejoiner_groups_buffer.go, line 17 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_wTEJ76_L5zaoHBINQ:-L_wTEJ76_L5zaoHBINR:b-947elx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c6a6f305bf9aac76fe6694501049efb3379f0f1/pkg/sql/exec/mergejoiner_groups_buffer.go#L17)):*
> ```Go
> package exec
> 
> type groupsBuffer struct {
> ```

Please add a comment here as well as to all the methods. I also think it might be good to describe how this struct buffers in the name, e.g. `circularGroupsBuffer`. 
___
*[pkg/sql/exec/mergejoiner_groups_buffer.go, line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_wSjv-3vilSnKP8YxW:-L_wSjv-3vilSnKP8YxX:bwj0ef9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c6a6f305bf9aac76fe6694501049efb3379f0f1/pkg/sql/exec/mergejoiner_groups_buffer.go#L29)):*
> ```Go
> }
> 
> func NewGroupsBuffer(bufferCap int) groupsBuffer {
> ```

nit: a constructor is only prefixed with `New` when it returns a pointer, this should be `MakeGroupsBuffer`. I also don't think this struct needs to be public.
___
*[pkg/sql/exec/mergejoiner_groups_buffer.go, line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_wVeNe8mIt9AuYWwFX:-L_wVeNe8mIt9AuYWwFY:blm890s) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c6a6f305bf9aac76fe6694501049efb3379f0f1/pkg/sql/exec/mergejoiner_groups_buffer.go#L92)):*
> ```Go
> 
> 	if endIdx < startIdx {
> 		copy(returnGroups, groups[startIdx:b.bufferCap])
> ```

Are these copies necessary? I don't have much context but it seems like you're copying into a parameter as well as returning it? This could definitely be something that affects the performance drop you've seen.
___
*[pkg/sql/exec/mergejoiner_test.go, line 490 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nQC3i9vJhmioqc6J8:-L_wXHcg0ABGBHBtdG6s:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner_test.go#L490)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Consider adding a randomized test as well with variable number of equality columns and small number of tuples. I think a randomized test I wrote for sort chunks might be a good starting point.
</blockquote></details>

+1


<!-- Sent from Reviewable.io -->

asubiotto reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35655)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/exec/mergejoiner.go, line 347 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_w_ftO2oaR_MelGX36:-L_w_ftO2oaR_MelGX37:ba4hut1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c6a6f305bf9aac76fe6694501049efb3379f0f1/pkg/sql/exec/mergejoiner.go#L347)):*
> ```Go
> 
> // buildSavedRuns expands the left and right runs in state into their cross product, dumping it
> // into the output buffer (and savedOutput overflow buffer if necessary).
> ```

I know this is from the previous PR but I think this definitely needs to be fixed. We can't save the whole cross product to the `savedOutput` and have it grow unbounded. If we stayed with this approach we would need to add disk spilling to the merge joiner which is a definite no-no. Instead, you should only build as many rows as ColBatchSize and return those. I think that having the merge joiner have an explicit state member (kind of like the sorter) will make this easier to implement and reason about. I'm fine tackling this in a separate PR, but just want to make sure that you're planning on doing this.


<!-- Sent from Reviewable.io -->

georgeutsin committed:
add randomized testing

georgeutsin committed:
fix equality column indexing in complete run

Release note: None

georgeutsin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35655)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/exec/mergejoiner.go, line 145 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nIof-1HQXvSo3WUsX:-L_nX3ROE7VZW6pZu46v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L145)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: it is no longer "a side".
</blockquote></details>

Done.
___
*[pkg/sql/exec/mergejoiner.go, line 241 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nKgIP08Ad02nSEbjV:-L_rN-Ou5hAXmSOQdoUm:b-ckmdru) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L241)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: I don't immediately see what `outCount` stands for and where it is coming from. Please extend the comment to explain it.
</blockquote></details>

Done.
Also refactored the function naming since the functionality has changed a little since its first inception.
___
*[pkg/sql/exec/mergejoiner.go, line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nKDhGB2sthDwLp-pj:-L_rNKDJAo0xo63OAOS8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L258)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: I think usually `takes` refers to the input argument, but the run comes from the state. I'd probably say something like `completeRun given the source input extends the existing run in state`.
</blockquote></details>

Done.
___
*[pkg/sql/exec/mergejoiner.go, line 265 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nJN301znhcmm2KwFL:-L_rNKmYAh1KVP4y72bV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L265)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: with this many return values consider naming them.
</blockquote></details>

Done.
___
*[pkg/sql/exec/mergejoiner.go, line 278 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nLf--C8LD6JX_ISdx:-L_rNNGL3QAfgKRx5Pq9:b-aehuz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L278)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: what does `minor` mean in this context?
</blockquote></details>

Done.
minor was a superfluous word 
___
*[pkg/sql/exec/mergejoiner.go, line 283 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nOcFS-65El9OBdgHv:-L_rNUfLDOdX8wKqw-fv:brm4rkr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L283)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think this is why you were asking about `tuplesDiffer`? If you want to use it, you could write something like (if `sel` is nil):
```
differ := false
for i := 0; i < len(source.eqCols)-1; i++ {
  colIdx := source.eqCols[i]
  tuplesDiffer(types.Int64, savedRun.ColVec(int(colIdx)), *savedRunIdx-1, bat.ColVec(int(colIdx)), rowIdx, &differ)
}
if differ {
  return rowIdx, bat, length, sel, keys
```
(But I guess you still need to template the rest, so it might not make sense to use it at the moment.)
</blockquote></details>

Yep, this is exactly what I was going to do, but I'm saving that for the templating PR.
___
*[pkg/sql/exec/mergejoiner.go, line 404 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nQiSREhvNWJjHqU46:-L_rOi_89zfALMQdinso:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L404)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: I think this phase deserves at least a high level comment explaining what we're doing here.
</blockquote></details>

Done.
___
*[pkg/sql/exec/mergejoiner.go, line 405 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nRGROCH6SDQV1VyGG:-L_wXWRfEWUwq8L7iSPd:b-6ld7vk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L405)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think you're using a circular buffer here, and we have an implementation of such already in `ring.Buffer`. Consider using that if possible.
</blockquote></details>

I rolled my own ADT here because we want to avoid using a generic interface in tight loops. Done.
___
*[pkg/sql/exec/mergejoiner.go, line 416 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nQzF1BPQTu-Cu2kZg:-L_rOyxT3XQdVkwvwZBV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner.go#L416)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: `in the column`.
</blockquote></details>

Done.
___
*[pkg/sql/exec/mergejoiner_groups_buffer.go, line 17 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_wTEJ76_L5zaoHBINQ:-LaG9ADWDP_wRhm9lm03:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c6a6f305bf9aac76fe6694501049efb3379f0f1/pkg/sql/exec/mergejoiner_groups_buffer.go#L17)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Please add a comment here as well as to all the methods. I also think it might be good to describe how this struct buffers in the name, e.g. `circularGroupsBuffer`. 
</blockquote></details>

Done.
___
*[pkg/sql/exec/mergejoiner_groups_buffer.go, line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_wVeNe8mIt9AuYWwFX:-LaGAHasEvys-OHMqZw9:b-5ywn7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c6a6f305bf9aac76fe6694501049efb3379f0f1/pkg/sql/exec/mergejoiner_groups_buffer.go#L92)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Are these copies necessary? I don't have much context but it seems like you're copying into a parameter as well as returning it? This could definitely be something that affects the performance drop you've seen.
</blockquote></details>

Done. I worked around this by allocating twice the memory needed for each circular buffer slice, so that making the elements contiguous is one operation.
___
*[pkg/sql/exec/mergejoiner_test.go, line 490 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nQC3i9vJhmioqc6J8:-LaGODbl1fIjUgvYyuR7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner_test.go#L490)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 148 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nN_es8fLipHZEP_e5:-LaGOfJy-qYNxAQg86KM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner_tmpl.go#L148)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: it'd be useful to explain the return values in the comment.
</blockquote></details>

Done.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-L_nNvyb8Behwuwg2nsO:-L_wXUKv6Gs3uG1CDveA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ea5683bc7fcdc95b84d27c508634a1d8bba7696/pkg/sql/exec/mergejoiner_tmpl.go#L150)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: this `outCount` seems to be not used anymore, maybe get rid off it? Or declare it within the for loop?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

georgeutsin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35655)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @georgeutsin, @jordanlewis, and @yuzefovich)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaGuVSICLNIKRUJbKH_:-LaGuVSICLNIKRUJbKHa:bl34e40) (no related file):*
RFAL :eyes:

---


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 1 of 4 files at r6, 4 of 4 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35655)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/exec/mergejoiner.go, line 268 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaGl5nP8fjOrrAEVKv6:-LaGl5nP8fjOrrAEVKv7:b-alzax8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95e2745c3049bd86d83c50155adf7912f994306/pkg/sql/exec/mergejoiner.go#L268)):*
> ```Go
> // completeRun extends the run in state given the source input.
> // To do this, we first check that the next batch is still the same run by checking all equality
> // columns except for the last minor equality column. Then we complete the run by Next'ing input
> ```

[nit]: `minor`?
___
*[pkg/sql/exec/mergejoiner.go, line 432 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaGqB5QEtZZ2NavYfkk:-LaGqB5QEtZZ2NavYfkl:b-aw7gvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95e2745c3049bd86d83c50155adf7912f994306/pkg/sql/exec/mergejoiner.go#L432)):*
> ```Go
> 				curLLength := lGroup.rowEndIdx
> 				curRLength := rGroup.rowEndIdx
> 				// Expand or filter each group based on the current equality column
> ```

[nit]: missing period.
___
*[pkg/sql/exec/mergejoiner_groups_buffer.go, line 30 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaGrVM91_SMEMB4v2KN:-LaGrVM91_SMEMB4v2KO:b-8lql8u) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95e2745c3049bd86d83c50155adf7912f994306/pkg/sql/exec/mergejoiner_groups_buffer.go#L30)):*
> ```Go
> }
> 
> func newGroupsBuffer(bufferCap int) circularGroupsBuffer {
> ```

[nit]: I think Alfonso mentioned that usually `new...` implies returning a pointer, so `makeGroupsBuffer` is probably more suitable.
___
*[pkg/sql/exec/mergejoiner_groups_buffer.go, line 53 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaGrybh6cXrXP4-bvgV:-LaGrybh6cXrXP4-bvgW:b-v5hzre) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95e2745c3049bd86d83c50155adf7912f994306/pkg/sql/exec/mergejoiner_groups_buffer.go#L53)):*
> ```Go
> 
> // nextGroupInCol returns whether or not there exists a next group in the current
> // column, and sets the parameters to be the left and right groups corresponding
> ```

[nit]: `corresponding to`.
___
*[pkg/sql/exec/mergejoiner_test.go, line 791 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaGughW76l1cwYVn3Sm:-LaGughW76l1cwYVn3Sn:bvogbqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95e2745c3049bd86d83c50155adf7912f994306/pkg/sql/exec/mergejoiner_test.go#L791)):*
> ```Go
> }
> 
> func TestMergeJoinerRandomized(t *testing.T) {
> ```

This is great, but having a variable number of equality columns will make it even greater :) I'm sure you have thought about this, and probably it is hard to generate such tests and especially the expected output, so maybe we could utilize existing merge join processor, i.e. generate some totally random input, sort it, and then pass it through both columnar merge join operator and row-based merge joiner processor?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35655)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/exec/mergejoiner.go, line 268 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaGl5nP8fjOrrAEVKv6:-LaGzYO42GJYqL3AqyKp:b-alzax8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95e2745c3049bd86d83c50155adf7912f994306/pkg/sql/exec/mergejoiner.go#L268)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: `minor`?
</blockquote></details>

[nit]: `minor`?


<!-- Sent from Reviewable.io -->

georgeutsin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35655)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/exec/mergejoiner.go, line 432 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaGqB5QEtZZ2NavYfkk:-LaGzyYQ-2DahoA3PzNP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95e2745c3049bd86d83c50155adf7912f994306/pkg/sql/exec/mergejoiner.go#L432)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: missing period.
</blockquote></details>

Done.
___
*[pkg/sql/exec/mergejoiner_groups_buffer.go, line 30 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaGrVM91_SMEMB4v2KN:-LaH-2ipEvLyjkfhkGXx:b-vkjwea) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95e2745c3049bd86d83c50155adf7912f994306/pkg/sql/exec/mergejoiner_groups_buffer.go#L30)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: I think Alfonso mentioned that usually `new...` implies returning a pointer, so `makeGroupsBuffer` is probably more suitable.
</blockquote></details>

Oops, that one snuck in there
___
*[pkg/sql/exec/mergejoiner_groups_buffer.go, line 53 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaGrybh6cXrXP4-bvgV:-LaH-6G2BkILWS3-93VK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95e2745c3049bd86d83c50155adf7912f994306/pkg/sql/exec/mergejoiner_groups_buffer.go#L53)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: `corresponding to`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
Could you share the benchmark differences with master by running `./scripts/bench`?

> Reviewed 2 of 2 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-:-LaNGvK5B1intsfZ7SNO:b2oqmj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/exec/mergejoiner.go, line 387 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaNKE0nAjxkbRtTjcy1:-LaNKE0nAjxkbRtTjcy2:b-mkmn89) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb5d58bd49cd9edebcdd026d27ab1bd89a8f14a/pkg/sql/exec/mergejoiner.go#L387)):*
> ```Go
> 
> 	for {
> 		eqColIdx := 0
> ```

It seems like you don't need the `eqColIdx` in this scope, I'd just remove this and use `:=` in the for loop below.
___
*[pkg/sql/exec/mergejoiner_groups_buffer.go, line 31 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaMfSsd9GMTjLuqYXM5:-LaMfSsd9GMTjLuqYXM6:b3kgp1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb5d58bd49cd9edebcdd026d27ab1bd89a8f14a/pkg/sql/exec/mergejoiner_groups_buffer.go#L31)):*
> ```Go
> 
> func makeGroupsBuffer(bufferCap int) circularGroupsBuffer {
> 	b := circularGroupsBuffer{
> ```

nit: you can `return circularGroupsBuffer` directly here
___
*[pkg/sql/exec/mergejoiner_groups_buffer.go, line 77 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaNFjkeAFQfJeWV3jiB:-LaNFjkeAFQfJeWV3jiC:b-r63cd6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb5d58bd49cd9edebcdd026d27ab1bd89a8f14a/pkg/sql/exec/mergejoiner_groups_buffer.go#L77)):*
> ```Go
> 	curLIdx int, lRunLength int, curRIdx int, rRunLength int,
> ) {
> 	b.leftGroups[b.bufferEndIdx] = group{curLIdx, curLIdx + lRunLength, rRunLength}
> ```

You could probably reuse the group that's already in this slot


<!-- Sent from Reviewable.io -->

georgeutsin commented:
@asubiotto I didn't include the benchmarks because the ones from master currently don't make sense, as they include a) the bug that skips some rows b) the benchmarks themselves are incorrect.

This PR includes fixes for the output shaping, so these benchmarks are more reliable. In any case, here are the results:

```
name                                      old time/op    new time/op    delta
MergeJoiner/rows=1024-8                     65.4µs ± 2%    69.6µs ± 1%      +6.38%  (p=0.000 n=9+10)
MergeJoiner/rows=4096-8                      130µs ± 1%     267µs ± 1%    +104.75%  (p=0.000 n=10+9)
MergeJoiner/rows=16384-8                     519µs ± 3%    1039µs ± 1%    +100.10%  (p=0.000 n=10+9)
MergeJoiner/rows=1048576-8                  31.9ms ± 4%    64.3ms ± 2%    +101.63%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=1024-8       66.6µs ± 2%    69.0µs ± 1%      +3.56%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=4096-8        133µs ± 1%     262µs ± 1%     +97.64%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=16384-8       515µs ± 2%    1047µs ± 3%    +103.49%  (p=0.000 n=10+9)
MergeJoiner/oneSideRepeat-rows=1048576-8    33.2ms ± 5%    65.8ms ± 2%     +98.55%  (p=0.000 n=10+9)
MergeJoiner/bothSidesRepeat-rows=1024-8     69.6µs ± 7%    70.2µs ± 2%        ~     (p=0.218 n=10+10)
MergeJoiner/bothSidesRepeat-rows=4096-8      624µs ± 9%    1356µs ± 1%    +117.27%  (p=0.000 n=9+8)
MergeJoiner/bothSidesRepeat-rows=16384-8    2.95ms ±17%   18.72ms ± 0%    +535.25%  (p=0.000 n=10+9)

name                                      old speed      new speed      delta
MergeJoiner/rows=1024-8                   1.00GB/s ± 2%  0.94GB/s ± 1%      -6.00%  (p=0.000 n=9+10)
MergeJoiner/rows=4096-8                   2.01GB/s ± 1%  0.98GB/s ± 1%     -51.16%  (p=0.000 n=10+9)
MergeJoiner/rows=16384-8                  2.02GB/s ± 3%  1.01GB/s ± 1%     -50.04%  (p=0.000 n=10+9)
MergeJoiner/rows=1048576-8                2.10GB/s ± 4%  1.04GB/s ± 2%     -50.42%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=1024-8      984MB/s ± 2%   950MB/s ± 1%      -3.44%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=4096-8     1.98GB/s ± 1%  1.00GB/s ± 1%     -49.40%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=16384-8    2.04GB/s ± 2%  1.00GB/s ± 3%     -50.85%  (p=0.000 n=10+9)
MergeJoiner/oneSideRepeat-rows=1048576-8  2.03GB/s ± 5%  1.02GB/s ± 2%     -49.67%  (p=0.000 n=10+9)
MergeJoiner/bothSidesRepeat-rows=1024-8    943MB/s ± 6%   934MB/s ± 2%        ~     (p=0.218 n=10+10)
MergeJoiner/bothSidesRepeat-rows=4096-8    413MB/s ±17%   193MB/s ± 1%     -53.20%  (p=0.000 n=10+8)
MergeJoiner/bothSidesRepeat-rows=16384-8   358MB/s ±15%    56MB/s ± 0%     -84.34%  (p=0.000 n=10+9)

name                                      old alloc/op   new alloc/op   delta
MergeJoiner/rows=1024-8                      12.0B ± 0%   4239.0B ± 0%  +35225.00%  (p=0.000 n=10+10)
MergeJoiner/rows=4096-8                      25.0B ± 0%   4285.0B ± 0%  +17040.00%  (p=0.000 n=10+10)
MergeJoiner/rows=16384-8                     86.0B ± 0%   4378.0B ± 0%   +4990.70%  (p=0.000 n=10+10)
MergeJoiner/rows=1048576-8                  5.20kB ± 0%   19.67kB ± 0%    +278.62%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=1024-8        12.0B ± 0%   4239.0B ± 0%  +35225.00%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=4096-8        25.0B ± 0%   4285.0B ± 0%  +17040.00%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=16384-8       86.0B ± 0%   4378.0B ± 0%   +4990.70%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=1048576-8    5.20kB ± 0%   19.67kB ± 0%    +278.62%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=1024-8      12.0B ± 0%   4239.0B ± 0%  +35225.00%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=4096-8      429kB ± 0%     433kB ± 0%      +1.06%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=16384-8    2.89MB ± 0%    2.90MB ± 0%      +0.24%  (p=0.000 n=9+9)

name                                      old allocs/op  new allocs/op  delta
MergeJoiner/rows=1024-8                       0.00           4.00 ± 0%       +Inf%  (p=0.000 n=10+10)
MergeJoiner/rows=4096-8                       0.00           4.00 ± 0%       +Inf%  (p=0.000 n=10+10)
MergeJoiner/rows=16384-8                      0.00           4.00 ± 0%       +Inf%  (p=0.000 n=10+10)
MergeJoiner/rows=1048576-8                    2.00 ± 0%      9.00 ± 0%    +350.00%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=1024-8         0.00           4.00 ± 0%       +Inf%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=4096-8         0.00           4.00 ± 0%       +Inf%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=16384-8        0.00           4.00 ± 0%       +Inf%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=1048576-8      2.00 ± 0%      9.00 ± 0%    +350.00%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=1024-8       0.00           4.00 ± 0%       +Inf%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=4096-8      5.19k ± 0%     5.20k ± 0%      +0.15%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=16384-8     30.9k ± 0%     30.9k ± 0%      +0.03%  (p=0.000 n=10+10)
```

georgeutsin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35655)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/exec/mergejoiner.go, line 387 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaNKE0nAjxkbRtTjcy1:-LaQghszAgPFweoJ-Fz0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb5d58bd49cd9edebcdd026d27ab1bd89a8f14a/pkg/sql/exec/mergejoiner.go#L387)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

It seems like you don't need the `eqColIdx` in this scope, I'd just remove this and use `:=` in the for loop below.
</blockquote></details>

Done.
___
*[pkg/sql/exec/mergejoiner_groups_buffer.go, line 77 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaNFjkeAFQfJeWV3jiB:-LaQgk2I39igCwbXfooO:ba302fy) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb5d58bd49cd9edebcdd026d27ab1bd89a8f14a/pkg/sql/exec/mergejoiner_groups_buffer.go#L77)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

You could probably reuse the group that's already in this slot
</blockquote></details>

When I edited the members of the group in the place, the benchmarks showed that was slower anyway so I'm keeping it as is


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35655)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/exec/mergejoiner_groups_buffer.go, line 77 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaNFjkeAFQfJeWV3jiB:-LaR7GJvBHnJDirJ7B_w:bwlngwa) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb5d58bd49cd9edebcdd026d27ab1bd89a8f14a/pkg/sql/exec/mergejoiner_groups_buffer.go#L77)):*
<details><summary><i>Previously, georgeutsin (George Utsin) wrote…</i></summary><blockquote>

When I edited the members of the group in the place, the benchmarks showed that was slower anyway so I'm keeping it as is
</blockquote></details>

That's surprising. I would expect it to be just noise. That's the nice thing about `./scripts/bench`, it runs your benchmarks 10 times and tosses out any outliers to get a statistically significant difference.
___
*[pkg/sql/exec/mergejoiner_test.go, line 776 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaR7XUo9ClP0fmyJSDz:-LaR7XUo9ClP0fmyJSE-:b4fpbdv) ([raw file](https://github.com/cockroachdb/cockroach/blob/72c1a1b1341bb094de963aab22470f68b5b3e09b/pkg/sql/exec/mergejoiner_test.go#L776)):*
> ```Go
> 	}
> 	//
> 	//// left repeats
> ```

I think you might have missed changing/removing this commented code in this commit.


<!-- Sent from Reviewable.io -->

georgeutsin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35655)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/exec/mergejoiner_test.go, line 776 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaR7XUo9ClP0fmyJSDz:-LaR97MX8VGNDeu2RzT7:bse8oi3) ([raw file](https://github.com/cockroachdb/cockroach/blob/72c1a1b1341bb094de963aab22470f68b5b3e09b/pkg/sql/exec/mergejoiner_test.go#L776)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think you might have missed changing/removing this commented code in this commit.
</blockquote></details>

This comment may be referencing an old revision


<!-- Sent from Reviewable.io -->

georgeutsin commented:
bors r+

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35655)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/exec/mergejoiner.go, line 268 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaRAXqD43fscNexoX77:-LaRAXqD43fscNexoX78:b-alzax8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a42acc7eb0cbe412ff899866df73d7caeca8fe1/pkg/sql/exec/mergejoiner.go#L268)):*
> ```Go
> // completeRun extends the run in state given the source input.
> // To do this, we first check that the next batch is still the same run by checking all equality
> // columns except for the last minor equality column. Then we complete the run by Next'ing input
> ```

[nit]: `minor`?
___
*[pkg/sql/exec/mergejoiner_test.go, line 791 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/35655#-LaGughW76l1cwYVn3Sm:-LaRA07N2cfksPFBbT1-:b-i15zlk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a95e2745c3049bd86d83c50155adf7912f994306/pkg/sql/exec/mergejoiner_test.go#L791)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

This is great, but having a variable number of equality columns will make it even greater :) I'm sure you have thought about this, and probably it is hard to generate such tests and especially the expected output, so maybe we could utilize existing merge join processor, i.e. generate some totally random input, sort it, and then pass it through both columnar merge join operator and row-based merge joiner processor?
</blockquote></details>

I plan on working on such test infrastructure this Friday, so we'll add such test later.


<!-- Sent from Reviewable.io -->

georgeutsin commented:
Flaky test, trying again:

bors r+
